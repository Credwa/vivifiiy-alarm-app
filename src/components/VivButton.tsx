import Colors from '@/constants/Colors';
import * as React from 'react';
import { StyleSheet, TouchableOpacity, Pressable, GestureResponderEvent, View } from 'react-native';
import { Icon } from '@expo/vector-icons/build/createIconSet';
import VivText from '@/components/VivText';

interface PaddingHorizontal {
  left: number;
  right: number;
}
interface ButtonProps {
  icon?: React.ReactComponentElement<Icon<string, any>>;
  text: string;
  style?: any;
  iconPosition?: 'left' | 'right';
  separator?: boolean;
  type?: 'basic' | 'separated';
  paddingHorizontal?: PaddingHorizontal;
  paddingVertical?: number;
  color?: 'Primary' | 'Secondary' | 'Default';
  onPress?: ((event: GestureResponderEvent) => void) | undefined;
}

export default function VivButton(props: ButtonProps) {
  let color = { default: Colors.white, pressed: Colors.greyLight1 };
  let margin = { marginLeft: 0, marginRight: 0 };
  switch (props.color) {
    case 'Primary':
      color = { default: Colors.blueLight2, pressed: Colors.blueLight3 };
      break;
    case 'Secondary':
      color = { default: Colors.orangeMedium, pressed: Colors.orangeMedium2 };
      break;
    default:
      color = color;
  }

  if (props.iconPosition === 'left') {
    margin.marginLeft = 8;
  } else if (props.iconPosition === 'right') {
    margin.marginRight = 8;
  }

  return !props.separator ? (
    <Pressable
      style={({ pressed }) => [
        {
          ...styles.button,
          ...props.style,
          backgroundColor: pressed ? color.pressed : color.default,
          paddingVertical: props.paddingVertical || 13.5,
          paddingLeft: props.paddingHorizontal && props.paddingHorizontal.left ? props.paddingHorizontal.left : 12.5,
          paddingRight: props.paddingHorizontal && props.paddingHorizontal.right ? props.paddingHorizontal.right : 12.5
        }
      ]}
      onPress={props.onPress}
    >
      {({ pressed }) => (
        <>
          {props.iconPosition && props.iconPosition === 'left' ? props.icon : null}
          <VivText fontName="Headline" color={pressed ? Colors.black : Colors.blueDark} style={margin}>
            {props.text}
          </VivText>
          {!props.iconPosition || props.iconPosition === 'right' ? props.icon : null}
        </>
      )}
    </Pressable>
  ) : (
    <Pressable
      style={({ pressed }) => [
        { ...props.style, ...styles.buttonSeparated, backgroundColor: pressed ? color.pressed : color.default }
      ]}
      onPress={props.onPress}
    >
      {({ pressed }) => (
        <>
          <View style={[styles.iconSeparatedView, { paddingVertical: props.paddingVertical || 13.5 }]}>
            {props.icon}
          </View>
          <VivText
            fontName="Headline"
            color={pressed ? Colors.black : Colors.blueDark}
            style={[
              styles.textSeparated,
              {
                paddingVertical: props.paddingVertical || 13.5,
                paddingLeft:
                  props.paddingHorizontal && props.paddingHorizontal.left ? props.paddingHorizontal.left : 12.5,
                paddingRight:
                  props.paddingHorizontal && props.paddingHorizontal.right ? props.paddingHorizontal.right : 12.5
              }
            ]}
          >
            {props.text}
          </VivText>
        </>
      )}
    </Pressable>
  );
}

const styles = StyleSheet.create({
  button: {
    flexDirection: 'row',
    textAlign: 'center',
    backgroundColor: '#FFFFFF',
    paddingHorizontal: 25,
    borderRadius: 7
  },
  buttonSeparated: {
    flexDirection: 'row',
    backgroundColor: '#FFFFFF',
    borderRadius: 7
  },
  iconSeparatedView: {
    borderRightWidth: 0.3,
    borderColor: Colors.greyLight2,
    paddingHorizontal: 18
  },
  textSeparated: {
    paddingVertical: 13.5,
    paddingLeft: 10,
    paddingRight: 25
  }
});
