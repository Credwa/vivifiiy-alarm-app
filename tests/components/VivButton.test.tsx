import VivButton from '@/components/VivButton';
import { Ionicons } from '@expo/vector-icons';
import * as Font from 'expo-font';
import React from 'react';
import renderer from 'react-test-renderer';

describe('<VivButton />', () => {
  beforeEach(async () => {
    await Font.loadAsync({
      ...Ionicons.font
    });
  });

  it(`renders a basic button correctly`, () => {
    const tree = renderer.create(<VivButton text="Hello World" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`renders a basic button correctly with all props`, () => {
    const tree = renderer
      .create(
        <VivButton
          text="Hello World props"
          color="Primary"
          onPress={() => {
            return;
          }}
          paddingVertical={14}
          paddingHorizontal={{ left: 12, right: 12 }}
          iconPosition="left"
          icon={<Ionicons name="md-alarm" size={22} />}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`renders a separated button correctly with all props`, () => {
    const tree = renderer
      .create(
        <VivButton
          text="Hello World props"
          color="Secondary"
          onPress={() => {
            return;
          }}
          separator
          paddingVertical={11}
          paddingHorizontal={{ left: 15, right: 15 }}
          icon={<Ionicons name="md-alarm" size={22} />}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
