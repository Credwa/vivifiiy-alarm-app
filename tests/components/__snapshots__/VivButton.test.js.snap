// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VivButton /> renders a basic button correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 7,
        "flexDirection": "row",
        "paddingHorizontal": 25,
      },
      Object {
        "backgroundColor": "#ffffff",
        "paddingLeft": 12.5,
        "paddingRight": 12.5,
        "paddingVertical": 13.5,
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#869DB2",
        },
        Object {
          "fontFamily": "Nunito_600SemiBold",
          "fontSize": 17,
        },
        Object {
          "color": "#191940",
        },
        Object {
          "marginLeft": 0,
          "marginRight": 8,
        },
      ]
    }
  >
    Hello World
  </Text>
</View>
`;

exports[`<VivButton /> renders a basic button correctly with all props 1`] = `
<View
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 7,
        "flexDirection": "row",
        "paddingHorizontal": 25,
      },
      Object {
        "backgroundColor": "#3DD7F2",
        "paddingLeft": 12,
        "paddingRight": 12,
        "paddingVertical": 14,
      },
    ]
  }
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 22,
        },
        undefined,
        Object {
          "fontFamily": "ionicons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#869DB2",
        },
        Object {
          "fontFamily": "Nunito_600SemiBold",
          "fontSize": 17,
        },
        Object {
          "color": "#191940",
        },
        Object {
          "marginLeft": 8,
          "marginRight": 0,
        },
      ]
    }
  >
    Hello World props
  </Text>
</View>
`;

exports[`<VivButton /> renders a separated button correctly with all props 1`] = `
<View
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 7,
        "flexDirection": "row",
      },
      Object {
        "backgroundColor": "#D98436",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "borderColor": "#869DB2",
          "borderRightWidth": 0.3,
          "paddingHorizontal": 18,
        },
        Object {
          "paddingVertical": 11,
        },
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 22,
          },
          undefined,
          Object {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#869DB2",
        },
        Object {
          "fontFamily": "Nunito_600SemiBold",
          "fontSize": 17,
        },
        Object {
          "color": "#191940",
        },
        Array [
          Object {
            "paddingLeft": 10,
            "paddingRight": 25,
            "paddingVertical": 13.5,
          },
          Object {
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingVertical": 11,
          },
        ],
      ]
    }
  >
    Hello World props
  </Text>
</View>
`;
